internal/k8s/reconciler/manager.go

```diff
@@ -177,23 +181,29 @@ func (m *GatewayReconcileManager) UpsertGateway(ctx context.Context, g *gwv1beta
 		return m.client.Update(ctx, g)
 	}
 
+	// Calling validate outside of the upsert process allows us to re-resolve any
+	// external references and set the statuses accordingly.
 	consulNamespace := m.consulNamespaceMapper(g.GetNamespace())
 
+	gwState := state.InitialGatewayState(g)
+	gwState.ConsulNamespace = consulNamespace
+
+	service := m.factory.deployer.Service(config, g)
+
+	// Validate Gateway and Listeners, writing state into gwState
+	err = m.gwValidator.Validate(ctx, gwState, g, service)
+	if err != nil {
+		return err
+	}
+
 	m.namespaceMap[utils.NamespacedName(g)] = consulNamespace
 	gateway := m.factory.NewGateway(NewGatewayConfig{
 		Gateway:         g,
+		State:           gwState,
 		Config:          config,
 		ConsulNamespace: consulNamespace,
 	})
 
-	// Calling validate outside of the upsert process allows us to re-resolve any
-	// external references and set the statuses accordingly. Since we actually
-	// have other object updates triggering reconciliation loops, this is necessary
-	// prior to dirty-checking on upsert.
-	if err := gateway.Validate(ctx); err != nil {
-		return err
-	}
-
```

Comment:
The general flow now is that we:
1. Create a default/initial `GatewayState`
2. Validate the `Gateway` which updates the `GatewayState`
3. Create our internal model - `K8sGateway` - which pairs the derived `GatewayState` with the `Gateway`